src/components/NewCommentForm.tsx

```diff
@@ -1,99 +1,188 @@
-import React from 'react';
+import React, { useState, ChangeEvent, FormEvent } from 'react';
+import { CommentData } from '../types/Comment';
+import cn from 'classnames';
+
+type Props = {
+  loader: boolean;
+  onAdd: (newComment: CommentData) => Promise<void>;
+};
+
+export const NewCommentForm: React.FC<Props> = ({ loader, onAdd }) => {
+  const [newComment, setNewComment] = useState<CommentData>({
+    name: '',
+    email: '',
+    body: '',
+  });
+  const [touched, setTouched] = useState({
+    name: false,
+    email: false,
+    body: false,
+  });
+
+  const handleInputChange = (
+    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
+  ) => {
+    const { name, value } = e.target;
+
+    setNewComment(prevComment => ({ ...prevComment, [name]: value }));
+    setTouched(prevTouched => ({ ...prevTouched, [name]: true }));
+  };
+
+  const handleReset = () => {
+    setNewComment({
+      name: '',
+      email: '',
+      body: '',
+    });
+    setTouched({
+      name: false,
+      email: false,
+      body: false,
+    });
+  };
+
+  const handleSubmit = async (e: FormEvent) => {
+    e.preventDefault();
+
+    const isNameEmpty = newComment.name === '';
+    const isEmailEmpty = newComment.email === '';
+    const isBodyEmpty = newComment.body === '';
+
+    if (isNameEmpty || isEmailEmpty || isBodyEmpty) {
+      setTouched({
+        name: isNameEmpty,
+        email: isEmailEmpty,
+        body: isBodyEmpty,
+      });
+
+      return;
+    }
+
+    await onAdd(newComment);
```

Comment:
The `onAdd` function is asynchronous, but its promise is not awaited. This could lead to unexpected behavior if the form is reset before the promise resolves. Consider using `await` to ensure that the form is not reset until the promise resolves.